memory_core/memory_core.py

```diff
@@ -0,0 +1,101 @@
+from common.configurable_model import ConfigurableModel
+import functools
+from bit_vector import BitVector
+from enum import Enum
+import magma as m
+import logging
+logging.basicConfig(level=logging.DEBUG)
+
+
+class Mode(Enum):
+    LINE_BUFFER = 0
+    FIFO = 1
+    SRAM = 2
+
+
+class Memory:
+    def __init__(self, address_width, data_width):
+        self.address_width = address_width
+        self.data_width = data_width
+        self.data_depth = 1 << address_width
+        self.memory = {BitVector(i, address_width): BitVector(0, data_width)
+                       for i in range(self.data_depth)}
+
+    def check_addr(fn):
+        @functools.wraps(fn)
+        def wrapped(self, addr, *args):
+            assert 0 <= addr < self.data_depth, \
+                f"Address ({addr}) must be within range(0, {self.data_depth})"
+            return fn(self, addr, *args)
+        return wrapped
+
+    @check_addr
+    def read(self, addr):
+        return self.memory[addr]
+
+    @check_addr
+    def write(self, addr, value):
+        if isinstance(value, BitVector):
+            assert value.num_bits <= self.data_width, \
+                f"value.num_bits must be <= {self.data_width}"
+        if isinstance(value, int):
+            assert value.bit_length() <= self.data_width, \
+                f"value.bit_length() must be <= {self.data_width}"
+        self.memory[addr] = value
+
+
+def gen_memory_core(data_width: int, data_depth: int):
+
+    CONFIG_ADDR = BitVector(0, 32)
+
+    class MemoryCore(ConfigurableModel(32, 32)):
```

Comment:
Ultimately read/write/read_and_write is not the "real" interface for the memory, only if it is in SRAM mode. When we incorporate more modes, we should think about what the top level functions should be, e.g. push/pop

Reply:
Yes, I think linebuffer mode will reveal this. Would the idea be to have one interface for all modes? Or enable different interfaces depending on the mode.

Reply:
I think there are "different" interfaces for each mode but all the methods are always available to call, they just just raise an error based on the mode.

I guess this is one place where the whole lift-the-port-level-interface vs. lower-the-user-level-interface question becomes muddy -- the ports really have different meaning based on the mode. But the mode is a configuration-time parameter (not a generation-time one).

Reply:
For the sake of getting this in, I vote we iterate after checking in.